apis/deprecated.go

```diff
@@ -0,0 +1,141 @@
+/*
+Copyright 2019 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package apis
+
+import (
+	"context"
+	"reflect"
+	"strings"
+)
+
+const (
+	deprecated = "Deprecated"
+)
+
+// CheckDeprecated checks whether the provided named deprecated fields
+// are set in a context where deprecation is disallowed.
+// This is a shallow check.
+func CheckDeprecated(ctx context.Context, obj interface{}) *FieldError {
+	return CheckDeprecatedUpdate(ctx, obj, nil)
+}
+
+// CheckDeprecated checks whether the provided named deprecated fields
+// are set in a context where deprecation is disallowed.
+// This is a shallow check.
+func CheckDeprecatedUpdate(ctx context.Context, obj interface{}, org interface{}) *FieldError {
+	if IsDeprecatedAllowed(ctx) {
+		return nil
+	}
+
+	var errs *FieldError
+	newFields := getPrefixedNamedFieldValues(deprecated, obj)
+
+	if nonZero(reflect.ValueOf(org)) {
+		orgFields := getPrefixedNamedFieldValues(deprecated, org)
+
+		for name, newValue := range newFields {
```

Comment:
This could use a comment elaborating on why it is sufficient to walk just `newFields` and not `orgFields`